diff --git a/quickassist/build_system/build_files/OS/linux_common_user_space_rules.mk b/quickassist/build_system/build_files/OS/linux_common_user_space_rules.mk
index d592201..0c1062b 100644
--- a/quickassist/build_system/build_files/OS/linux_common_user_space_rules.mk
+++ b/quickassist/build_system/build_files/OS/linux_common_user_space_rules.mk
@@ -68,7 +68,7 @@ $(OBJECTS): %.o: %.c
 # Create the shared library
 $(LIB_SHARED): obj
 	@cd $($(PROG_ACY)_FINAL_OUTPUT_DIR); \
-	$(LINKER) $(LIB_SHARED_FLAGS) -o $@  $(OBJECTS) $(ADDITIONAL_OBJECTS) -lc
+	($(LINKER) $(LIB_SHARED_FLAGS) -o $@  $(OBJECTS) $(ADDITIONAL_OBJECTS) || true)
 
 
 # Create the static library
diff --git a/quickassist/lookaside/access_layer/src/Makefile b/quickassist/lookaside/access_layer/src/Makefile
index e4a4f2c..fae3553 100644
--- a/quickassist/lookaside/access_layer/src/Makefile
+++ b/quickassist/lookaside/access_layer/src/Makefile
@@ -217,9 +217,9 @@ lib_shared: lib
 	done ;\
 	cd $(ICP_FINAL_OUTPUT_DIR);\
 	cmd="$(LINKER) $(LIB_SHARED_FLAGS) -o $(LIB_SHARED) $(ADDITIONAL_OBJECTS) $(ADDITIONAL_LIBS) *.o -lpthread -ludev -lz\
-		-Bstatic -L$(ADF_DIR)/src -ladf_user -L$(OSAL_DIR)/src/build/$(ICP_OS)/$(ICP_OS_LEVEL)/ -losal -Bdynamic -lc"; \
+		-Bstatic -L$(ADF_DIR)/src -ladf_user -L$(OSAL_DIR)/src/build/$(ICP_OS)/$(ICP_OS_LEVEL)/ -losal -Bdynamic"; \
 	echo "$$cmd"; \
-	$$cmd
+	($$cmd || true)
 
 endif
 
